src/Wallet.purs

```diff
@@ -32,21 +43,95 @@ import Effect.Exception (throw)
 import Effect.Ref as Ref
 import FfiHelpers (MaybeFfiHelper, maybeFfiHelper)
 import Helpers ((<<>>))
+import Serialization (publicKeyFromPrivateKey, publicKeyHash)
 import Serialization as Serialization
-import Serialization.Address (Address, addressFromBytes)
-import Types.ByteArray (byteArrayToHex)
-import Types.CborBytes
-  ( CborBytes
-  , cborBytesToHex
-  , rawBytesAsCborBytes
+import Serialization.Address
+  ( Address
+  , NetworkId
+  , addressFromBytes
+  , enterpriseAddress
+  , enterpriseAddressToAddress
+  , keyHashCredential
   )
+import Serialization.Types (PrivateKey)
+import Types.ByteArray (byteArrayToHex)
+import Types.CborBytes (CborBytes, cborBytesToHex, rawBytesAsCborBytes)
 import Types.RawBytes (RawBytes, hexToRawBytes)
 import Types.Transaction (TransactionHash(TransactionHash))
 import Untagged.Union (asOneOf)
 
--- At the moment, we only support Nami's wallet. In the future we will expand
--- this with more constructors to represent out-of-browser wallets (e.g. WBE)
-data Wallet = Nami NamiWallet
+data Wallet
+  = Nami NamiWallet
+  | KeyWallet KeyWallet'
+
+-------------------------------------------------------------------------------
+-- Key backend
+-------------------------------------------------------------------------------
+type KeyWallet' =
+  { address :: NetworkId -> Aff Address
+  , selectCollateral :: Utxo -> Maybe TransactionUnspentOutput
+  , signTx :: Transaction -> Aff Transaction
+  }
+
+mkKeyWallet :: PrivateKey -> Wallet
+mkKeyWallet key =
+  KeyWallet
+    { address
+    , selectCollateral
+    , signTx
+    }
+  where
+  address :: NetworkId -> Aff Address
+  address network = publicKeyFromPrivateKey key
+    # liftEffect
+    <#> publicKeyHash
+      >>> keyHashCredential
+      >>> { network, paymentCred: _ }
+      >>> enterpriseAddress
+      >>> enterpriseAddressToAddress
+
+  selectCollateral :: Utxo -> Maybe TransactionUnspentOutput
+  selectCollateral utxos = unwrap <<< unwrap <$> flip
+    foldMapWithIndex
+    utxos
+    \input output ->
+      let
+        txuo = AdaOut $ TransactionUnspentOutput { input, output }
+        Value ada (NonAdaAsset naa) = _value txuo
+        onlyAda = all (all ((==) zero)) naa
+        bigAda = ada >= mkCoin 5_000_000
```

Comment:
Is hardcoding 5 ADA value safe?

Reply:
Not sure, but given Nami has the same collateral it's probably safe. Currently collateral is 150% of the tx fee, and I don't think fees can get that high?

Reply:
But also, this method can be overridden if necessary for big transactions

Reply:
No I mean is it determined by protocol parameters? We need to support private testnets too.

Reply:
Oh true. Once again it can be overridden, alternatively we can make it a parameter mkKeyWallet

Reply:
We could try and supply a collateral estimate to this function based on an estimated fee and scaling by collateral scale?

Reply:
Let's ignore this for now, I will create an issue to investigate.